pkg/generator/statemanager/statemanager.go

```diff
@@ -0,0 +1,227 @@
+	}
+	return nil
+}
+
+func gcGeneratorStateKey(entry gc.Entry, key string) string {
+	return fmt.Sprintf("[%s]-%s", key, entry.Key())
+}
+
+// SetLatest sets the latest state for the given key.
+// It will commit the state on success or move the state to GC on failure.
+func (m *Manager) SetLatest(ctx context.Context, kubeClient client.Client, stateKey, namespace string, resource *apiextensions.JSON, gen v1alpha1.Generator, state v1alpha1.GeneratorProviderState) {
+	m.internalState = append(m.internalState, QueueItem{
+		// Store state at .Latest[<key>] on success
+		// or attempt to immediately delete the state on failure
+		Commit: func() error {
+			genState := m.resource.GetGeneratorState()
+			if genState.Latest == nil {
+				genState.Latest = make(map[string]*v1beta1.GeneratorResourceState)
+			}
+			genState.Latest[stateKey] = &v1beta1.GeneratorResourceState{
+				Resource: resource,
+				State:    state,
+			}
+			return nil
+		},
+		// Rollback by cleaning up the state.
+		// In case of failure, move the state to GC so it will be cleaned up later.
+		Rollback: func() error {
+			err := gen.Cleanup(ctx, resource, state, kubeClient, namespace)
+			if err == nil {
+				return nil
+			}
+			return m.ImmediateMoveStateToGC(resource, fmt.Sprintf("rollback-%s", uuid.New().String()), gen, state)
+		},
+	})
+}
+
+// GetLatest returns the latest state for the given key.
+func (m *Manager) GetLatest(key string) *apiextensions.JSON {
+	state := m.resource.GetGeneratorState()
+	if state.Latest == nil {
+		return nil
+	}
+	latest := state.Latest[key]
+	if latest == nil {
+		return nil
+	}
+	return latest.State
+}
+
+// CleanupImmediate will cleanup the generator state immediately.
+// This is useful when we want to cleanup the state immediately after deletion of the resource.
+func CleanupImmediate(ctx context.Context, resource v1beta1.GeneratorStateManagingResource, kubeClient client.Client, namespace string) error {
+	var errs []error
+	generatorState := resource.GetGeneratorState()
+	for _, gcState := range generatorState.GC {
+		genImpl, err := v1alpha1.GetGenerator(gcState.Resource)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("unable to get generator: %w", err))
+			continue
+		}
+		err = genImpl.Cleanup(ctx, gcState.Resource, gcState.State, kubeClient, namespace)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("failed to cleanup generator state: %w", err))
+		}
+	}
+	return errors.Join(errs...)
+}
+
+// GarbageCollect will cleanup the generator state that is flagged for GC.
+// It updates the generator state with the new GC state.
+// Garbage collection is done in a best-effort manner, i.e. if an error occurs during cleanup of a generator state,
+// it will be aggregated and returned at the end but the cleanup will continue for the remaining generator states.
+func GarbageCollect(ctx context.Context, resource v1beta1.GeneratorStateManagingResource, kubeClient client.Client, namespace string) error {
```

Comment:
On cleanup failure should we set some kind of cleanup error state for the GC so it requeues based on that error and tries again?

Reply:
The garbage collection happens in two places:
1. when the reconcile() has finished, just before marking it as complete
2. when deletion of ExternalSecret/PushSecret is requested

If garbage collection fails it is marked as failed and is re-queued with exponential backoff. Is that good enough?

Reply:
Oh yeah. I wasn't sure if there is a re-queue. I had to draw the flow. :D Thanks.